### PR TITLE
Minimaldata from genesis

### PR DESCRIPTION
**Pull request by EyeOfPython**
- Remove `SCRIPT_VERIFY_MINIMALDATA`.
- Set `fRequireMinimal` in `interpreter.cpp` always to `true`. It is kept to not make the diff unnecessarily large.
- Fix minimal pushes in `src/test/data/script_tests.json`.
  - Remove tests that explicitly test for non-minimalness.
- Remove tests in `src/test/data/tx_valid.json` that explicitly test for non-minimalness.

### CODE REVIEW

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -220,7 +220,7 @@ bool EvalScript(std::vector<valtype> &stack, const CScript &script,
         return set_error(serror, ScriptError::SCRIPT_SIZE);
     }
     int nOpCount = 0;
-    bool fRequireMinimal = (flags & SCRIPT_VERIFY_MINIMALDATA) != 0;
```
Comment (schancel) [inline src/script/interpreter.cpp:223]: This seems wise. Maybe add a TODO or something here for posterity to go and delete the excess tests and branches.

**Review by schancel — APPROVED** [review]

File: src/script/interpreter.cpp (hunk 1 above)
Comment (EyeOfPython) [reply src/script/interpreter.cpp:223]: How about moving it into some constant somewhere and changing the case to SCREAM_SNAKE?

File: src/test/data/script_tests.json [hunk 2]
```diff
@@ -20,18 +20,20 @@
...
+ "1", "P2SH,STRICTENC", "MINIMALDATA"],
+["PUSHDATA2 0x0001 0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
+ "SIZE 0x020001 EQUAL", "P2SH,STRICTENC", "OK"],
 
-["PUSHDATA1 0x00","0 EQUAL", "P2SH,STRICTENC", "OK"],
-["PUSHDATA2 0x0000","0 EQUAL", "P2SH,STRICTENC", "OK"],
-["PUSHDATA4 0x00000000","0 EQUAL", "P2SH,STRICTENC", "OK"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:34]: For context; I removed these non-minimal pushes and replaced them with equivalent minimal pushes (or test for `MINIMALDATA` just at the boundary of becoming invalid)

File: src/test/data/script_tests.json [hunk 3]
```diff
@@ -168,18 +170,9 @@
...
-["0x01 0x81", "DUP BOOLOR", "P2SH,STRICTENC", "OK", "-1 -1 BOOLOR"],
-["0x01 0x80", "DUP BOOLAND", "P2SH,STRICTENC", "EVAL_FALSE", "negative-0 negative-0 BOOLAND"],
-["0x01 0x00", "DUP BOOLAND", "P2SH,STRICTENC", "EVAL_FALSE", " non-minimal-0  non-minimal-0 BOOLAND"],
-["0x01 0x81", "DUP BOOLAND", "P2SH,STRICTENC", "OK", "-1 -1 BOOLAND"],
-["0x01 0x00", "NOT", "P2SH,STRICTENC", "OK", "non-minimal-0 NOT"],
-["0x01 0x80", "NOT", "P2SH,STRICTENC", "OK", "negative-0 NOT"],
-["0x01 0x81", "NOT", "P2SH,STRICTENC", "EVAL_FALSE", "negative 1 NOT"],
-["0x01 0x80 0", "NUMEQUAL", "P2SH", "OK", "-0 0 NUMEQUAL"],
-["0x01 0x00 0", "NUMEQUAL", "P2SH", "OK", "non-minimal-0 0 NUMEQUAL"],
-["0x02 0x0000 0", "NUMEQUAL", "P2SH", "OK", "non-minimal-0 0 NUMEQUAL"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:182]: For context; there is no non-minimal-0 anymore, so this test makes no sense anymore.

File: src/test/data/script_tests.json [hunk 4]
```diff
@@ -238,10 +231,9 @@
 ["''", "DUP HASH256 SWAP SHA256 SHA256 EQUAL", "P2SH,STRICTENC", "OK"],
 ["''", "NOP HASH160 0x14 0xb472a266d0bd89c13706a4132ccfb16f7c3b9fcb EQUAL", "P2SH,STRICTENC", "OK"],
 ["'a'", "HASH160 NOP 0x14 0x994355199e516ff76c4fa4aab39337b9d84cf12b EQUAL", "P2SH,STRICTENC", "OK"],
-["'abcdefghijklmnopqrstuvwxyz'", "HASH160 PUSHDATA1 0x14 0xc286a1af0947f58d1ad787385b1c2c4a976f9e71 EQUAL", "P2SH,STRICTENC", "OK"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:241]: Tests for non-minimal "P2SH", which actually is no P2SH and just an ordinary script. This is invalid now; there can be no non-minimal "P2SH", therefore removed.

File: src/test/data/script_tests.json [hunk 5]
```diff
@@ -381,16 +372,6 @@
 ["2147483647", "1ADD 1", "P2SH,STRICTENC", "OK"],
 ["-2147483647", "1ADD 1", "P2SH,STRICTENC", "OK"],
 
-["1", "0x02 0x0100 EQUAL NOT", "P2SH,STRICTENC", "OK", "Not the same byte array..."],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:384]: For context; this test for non-minimal number pushes which are illegal now.

File: src/test/data/script_tests.json [hunk 6]
```diff
@@ -528,198 +509,116 @@
...
-["PUSHDATA2 0xFF00 0x111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111", "PUSHDATA1 0xFF 0x111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111 EQUAL", "", "OK", "PUSHDATA2 of 255 bytes equals PUSHDATA1 of it"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:546]: For context; all PUSHDATA forms are no longer equivalent (you can only ever use one form depending on size), so I removed the tests that test for that.

File: src/test/data/script_tests.json [hunk 7]
```diff
@@ -528,198 +509,116 @@
...
-["0x01 0x0e", "0x5e EQUAL", "", "OK", "OP_14 pushes 0x0e"],
-["0x01 0x0f", "0x5f EQUAL", "", "OK", "OP_15 pushes 0x0f"],
-["0x01 0x10", "0x60 EQUAL", "", "OK", "OP_16 pushes 0x10"],
-
-["Equivalency of different numeric encodings"],
-
-["0x02 0x8000", "128 NUMEQUAL", "", "OK", "0x8000 equals 128"],
-["0x01 0x00", "0 NUMEQUAL", "", "OK", "0x00 numequals 0"],
-["0x01 0x80", "0 NUMEQUAL", "", "OK", "0x80 (negative zero) numequals 0"],
-["0x02 0x0080", "0 NUMEQUAL", "", "OK", "0x0080 numequals 0"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:575]: For context; there is no "different numeric encodings" anymore.

File: src/test/data/script_tests.json [hunk 8]
```diff
@@ -528,198 +509,116 @@
...
+["0 IF 0x01 0x0c ENDIF 1", "", "", "OK", "OP_12 equiv"],
+["", "0 IF 0x01 0x0c ENDIF 1", "", "OK", "OP_12 equiv (scriptPubKey)"],
+["0 IF 0x01 0x0d ENDIF 1", "", "", "OK", "OP_13 equiv"],
+["", "0 IF 0x01 0x0d ENDIF 1", "", "OK", "OP_13 equiv (scriptPubKey)"],
+["0 IF 0x01 0x0e ENDIF 1", "", "", "OK", "OP_14 equiv"],
+["", "0 IF 0x01 0x0e ENDIF 1", "", "OK", "OP_14 equiv (scriptPubKey)"],
+["0 IF 0x01 0x0f ENDIF 1", "", "", "OK", "OP_15 equiv"],
+["", "0 IF 0x01 0x0f ENDIF 1", "", "OK", "OP_15 equiv (scriptPubKey)"],
+["0 IF 0x01 0x10 ENDIF 1", "", "", "OK", "OP_16 equiv"],
+["", "0 IF 0x01 0x10 ENDIF 1", "", "OK", "OP_16 equiv (scriptPubKey)"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:577]: For context; removed the now deleted "MINIMALDATA" script flag

File: src/test/data/script_tests.json [hunk 9]
```diff
@@ -528,198 +509,116 @@
...
+["0x02 0x0c00", "1", "", "OK"],
+["", "0x02 0x0c00 1", "", "OK"],
+["0x02 0x0d00", "1", "", "OK"],
+["", "0x02 0x0d00 1", "", "OK"],
+["0x02 0x0e00", "1", "", "OK"],
+["", "0x02 0x0e00 1", "", "OK"],
+["0x02 0x0f00", "1", "", "OK"],
+["", "0x02 0x0f00 1", "", "OK"],
+["0x02 0x1000", "1", "", "OK"],
+["", "0x02 0x1000 1", "", "OK"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:618]: For context; again tests that test for non-minimal pushes.

File: src/test/data/script_tests.json [hunk 10]
```diff
@@ -1020,7 +919,6 @@
 ["0x10 0x0102030405060708090A0B0C0D0E0F10 DUP CAT DUP CAT DUP CAT DUP CAT DUP CAT 0x08 0x0102030405060708 CAT 520", "NUM2BIN 0x10 0x0102030405060708090A0B0C0D0E0F10 DUP CAT DUP CAT DUP CAT DUP CAT DUP CAT 0x08 0x0102030405060708 CAT EQUAL", "P2SH,STRICTENC", "OK", "520 byte 1st operand"],
 ["0x05 0x0000000000 5", "NUM2BIN 0x05 0x0000000000 EQUAL", "P2SH,STRICTENC", "OK", "1st operand not minimally encoded"],
 ["0x05 0x0001000000 3", "NUM2BIN 0x03 0x000100 EQUAL", "P2SH,STRICTENC", "OK", "1st operand can shrink"],
-["1 0x03 0x050000", "NUM2BIN", "P2SH,STRICTENC", "OK", "2nd operand not minimally encoded"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:1023]: For context; again tests that test for non-minimal pushes.

File: src/test/data/script_tests.json [hunk 11]
```diff
@@ -1516,148 +1414,148 @@
...
+["0x01 0x0d", "DROP 1", "", "MINIMALDATA"],
+["", "0x01 0x0d DROP 1", "", "MINIMALDATA"],
+["0x01 0x0e", "DROP 1", "", "MINIMALDATA"],
+["", "0x01 0x0e DROP 1", "", "MINIMALDATA"],
+["0x01 0x0f", "DROP 1", "", "MINIMALDATA"],
+["", "0x01 0x0f DROP 1", "", "MINIMALDATA"],
+["0x01 0x10", "DROP 1", "", "MINIMALDATA"],
+["", "0x01 0x10 DROP 1", "", "MINIMALDATA"],
+
+["PUSHDATA1 0x48 0x111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111", "DROP 1", "",
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:1454]: For context; removed the now deleted MINIMALDATA script flag.

File: src/test/data/script_tests.json [hunk 12]
```diff
@@ -1516,148 +1414,148 @@
...
  "PUSHDATA4 of 256 bytes minimally represented by PUSHDATA2"],
-["", "PUSHDATA4 0x00010000 0x11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111 DROP 1", "MINIMALDATA",
+["", "PUSHDATA4 0x00010000 0x11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111 DROP 1", "",
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:1471]: For context; removed the now deleted MINIMALDATA script flag.

File: src/test/data/script_tests.json [hunk 13]
```diff
@@ -1516,148 +1414,148 @@
...
-["0x02 0x0181", "HASH160 0x14 0x823ceb939791c0227262f8c2e5d29f072a590609 OP_EQUAL", "MINIMALDATA", "OK", "non-minimal push inside P2SH redeemScript but P2SH not active"],
+["0x01 0x4f", "HASH160 0x14 0xc692a0d72bb4690b00f2728775fb8e45eeb190ad OP_EQUAL", "P2SH", "OK", "minimal push inside P2SH redeemScript"],
+["PUSHDATA1 0x01 0x4f", "HASH160 0x14 0xc692a0d72bb4690b00f2728775fb8e45eeb190ad OP_EQUAL", "P2SH", "MINIMALDATA", "P2SH redeemScript itself was not pushed minimally"],
+["1 0x01 0x81", "HASH160 0x14 0x19e672940307de32037c538ad37578ab38f2b83e OP_EQUAL", "P2SH", "MINIMALDATA", "P2SH redeemscript 'OP_BIN2NUM' pushed naively"],
+["1 -1", "HASH160 0x14 0x19e672940307de32037c538ad37578ab38f2b83e OP_EQUAL", "P2SH", "OK", "P2SH redeemscript 'OP_BIN2NUM' pushed minimally"],
+["0x02 0x0181", "HASH160 0x14 0x823ceb939791c0227262f8c2e5d29f072a590609 OP_EQUAL", "P2SH", "MINIMALDATA", "non-minimal push inside P2SH redeemScript"],
+["0x02 0x0181", "HASH160 0x14 0x823ceb939791c0227262f8c2e5d29f072a590609 OP_EQUAL", "", "OK", "non-minimal push inside P2SH redeemScript but P2SH not active"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:1481]: For context; removed the now deleted MINIMALDATA script flag.

File: src/test/data/script_tests.json [hunk 14]
```diff
@@ -1516,148 +1414,148 @@
...
+["0x01 0x80", "NOT DROP 1", "", "UNKNOWN_ERROR", "0x80 (negative zero) numequals 0"],
+["0x02 0x0080", "NOT DROP 1", "", "UNKNOWN_ERROR", "numequals 0"],
+["0x02 0x0500", "NOT DROP 1", "", "UNKNOWN_ERROR", "numequals 5"],
+["0x03 0x050000", "NOT DROP 1", "", "UNKNOWN_ERROR", "numequals 5"],
+["0x02 0x0580", "NOT DROP 1", "", "UNKNOWN_ERROR", "numequals -5"],
+["0x03 0x050080", "NOT DROP 1", "", "UNKNOWN_ERROR", "numequals -5"],
+["0x03 0xff7f80", "NOT DROP 1", "", "UNKNOWN_ERROR", "Minimal encoding is 0xffff"],
+["0x03 0xff7f00", "NOT DROP 1", "", "UNKNOWN_ERROR", "Minimal encoding is 0xff7f"],
+["0x04 0xffff7f80", "NOT DROP 1", "", "UNKNOWN_ERROR", "Minimal encoding is 0xffffff"],
+["0x04 0xffff7f00", "NOT DROP 1", "", "UNKNOWN_ERROR", "Minimal encoding is 0xffff7f"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:1496]: For context; removed the now deleted MINIMALDATA script flag.

File: src/test/data/script_tests.json [hunk 15]
```diff
@@ -1516,148 +1414,148 @@
...
+
+["0x02 0x0000 1", "SPLIT 2DROP 1", "", "OK", "first operand isn't a number, so no encoding rules"],
+["0x02 0x0000 0x02 0x0100", "SPLIT 2DROP 1", "", "UNKNOWN_ERROR"],
+["0x16 0x11111111111111111111111111111111111111110000 0x01 0x14", "NUM2BIN DROP 1", "", "OK", "first operand is an unlimited number and exempted from length/minimal encoding rules"],
+["0x16 0x11111111111111111111111111111111111111110000 0x02 0x1400", "NUM2BIN DROP 1", "", "UNKNOWN_ERROR"],
+
+["0x16 0xffffff7f000000000000000000000000000000000000", "BIN2NUM DROP 1", "", "OK", "first operand is a number limited in magnitude but exempted from length/minimal encoding rules"],
+
+["0x02 0x0000", "CHECKLOCKTIMEVERIFY DROP 1", "CHECKLOCKTIMEVERIFY", "UNKNOWN_ERROR"],
+["0x02 0x0000", "CHECKSEQUENCEVERIFY DROP 1", "CHECKSEQUENCEVERIFY", "UNKNOWN_ERROR"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:1558]: For context; removed the now deleted MINIMALDATA script flag.

File: src/test/data/script_tests.json [hunk 16]
```diff
@@ -1856,7 +1754,7 @@
     "0x05 0x004c0245aa",
     "HASH160 0x14 0xd3ec673296c7fd7e1a9e53bfc36f414de303e905 EQUAL",
     "CLEANSTACK,P2SH",
-    "CLEANSTACK",
+    "MINIMALDATA",
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:1757]: For context; the spirit of this check is to validate an invalid witness program; now the error is simply a bit earlier.

File: src/test/data/script_tests.json [hunk 17]
```diff
@@ -3041,428 +2939,428 @@
...
-    "INPUT_SIGCHECKS,MINIMALDATA,NULLFAIL,P2SH,SCHNORR_MULTISIG,STRICTENC",
+    "INPUT_SIGCHECKS,NULLFAIL,P2SH,SCHNORR_MULTISIG,STRICTENC",
     "OK",
     "Very short P2SH multisig 0-of-10, spent with schnorr mode (0 sigchecks)"
 ],
 
 ["CHECKSEQUENCEVERIFY tests"],
 ["", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "INVALID_STACK_OPERATION", "CSV automatically fails on an empty stack"],
 ["-1", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "NEGATIVE_LOCKTIME", "CSV automatically fails if stack top is negative"],
-["0x0100", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY,MINIMALDATA", "UNKNOWN_ERROR", "CSV fails if stack top is not minimally encoded"],
```
Comment (EyeOfPython) [inline src/test/data/script_tests.json:3465]: For context; removed the now deleted MINIMALDATA script flag.

File: src/test/data/tx_valid.json [hunk 18]
```diff
@@ -222,10 +222,6 @@
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "2147483647 2147483647 ADD CHECKLOCKTIMEVERIFY 1"]],
 "0100000001000100000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000feffffff", "P2SH,CHECKLOCKTIMEVERIFY"],
 
-["5 byte non-minimally-encoded arguments are valid"],
```
Comment (EyeOfPython) [inline src/test/data/tx_valid.json:225]: For context; they are not.

File: src/test/data/tx_valid.json [hunk 19]
```diff
@@ -298,10 +294,6 @@
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0x050000008001 CHECKSEQUENCEVERIFY 1"]],
 "020000000100010000000000000000000000000000000000000000000000000000000000000000000000ffffffff0100000000000000000000000000", "P2SH,CHECKSEQUENCEVERIFY"],
 
-["5 byte non-minimally-encoded operandss are valid"],
-[[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0x05 0x0000000000 CHECKSEQUENCEVERIFY 1"]],
```
Comment (EyeOfPython) [inline src/test/data/tx_valid.json:302]: For context; they are not.

File: src/test/monolith_opcodes_tests.cpp [hunk 20]
```diff
@@ -657,43 +657,23 @@ static void CheckDivMod(const valtype &a, const valtype &b,
...
+                    ScriptError::UNKNOWN);
+
+        CheckError(flags, {b, {0x00}}, CScript() << OP_DIV,
+                    ScriptError::UNKNOWN);
+        CheckError(flags, {b, {0x80}}, CScript() << OP_DIV,
+                    ScriptError::UNKNOWN);
+        CheckError(flags, {b, {0x00, 0x00}}, CScript() << OP_DIV,
+                    ScriptError::UNKNOWN);
+        CheckError(flags, {b, {0x00, 0x80}}, CScript() << OP_DIV,
+                    ScriptError::UNKNOWN);
```
Comment (EyeOfPython) [inline src/test/monolith_opcodes_tests.cpp:676]: `SCRIPT_VERIFY_MINIMALDATA` is always enabled (and thus removed from the code).

File: src/test/script_tests.cpp [hunk 21]
```diff
@@ -2445,40 +2445,7 @@ BOOST_AUTO_TEST_CASE(script_json_test) {
...
-                           SCRIPT_VERIFY_P2SH, BaseSignatureChecker(), &err));
-    BOOST_CHECK(pushdata2Stack == directStack);
-    BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
-
-    std::vector<std::vector<uint8_t>> pushdata4Stack;
-    BOOST_CHECK(EvalScript(pushdata4Stack,
-                           CScript(pushdata4, pushdata4 + sizeof(pushdata4)),
-                           SCRIPT_VERIFY_P2SH, BaseSignatureChecker(), &err));
-    BOOST_CHECK(pushdata4Stack == directStack);
-    BOOST_CHECK_MESSAGE(err == ScriptError::OK, ScriptErrorString(err));
```
Comment (EyeOfPython) [inline src/test/script_tests.cpp:2481]: For context; these tests make no sense anymore, as there's only one single way to push data.

**Pull request merged**